README.md
# Project 2 - ZETA: Network sniffer

## Zadání:
- Navrhněte a implementujte síťový analyzátor, který bude schopen zachytit a filtrovat pakety na konkrétním síťovém rozhrání
- Vytvořte dokumentaci k projektu.

## Úvod:
Program implementovaný v jazyce C, který zachycuje a analyzuje síťové pakety s využitím knihovny libpcap. Tato knihovna nabízí funkce pro zachycení síťového provozu a analýzu packetů. Celá implementace se nachází v souboru `ipk-sniffer.c`.

## Implementace ipk-sniffer.c:
Na záčátku ve funkci `main()` se použije funkce `pcap_findalldevs()`, která načte všechny dostupné rozhrání a vrátí ukazatel na první rozhrání v seznamě. Náseledně se zkontroluje zda je počet argumentů větší jak dva (pokud je dán jeden argument jsou vytisknuté dostupné rozhraní pomocí funkce `print_interface()`, která na konci uvolní seznam alokovaných zařízení pomocí funkce `pcap_freealldevs()`) a zahájí se cyklus na kontrolu argumentů a nastavení vlajek přítomnosti protokolů. Tato kontrola probíhá pomocí funkce `getopt()` z knihovny `getopt.h`, která postupně parsuje argumenty. V rámci kontroli `inteface` se kontroluje zda se za argumentem nachází název a ne další argument.Tato kontrola podporuje krátké (např. `-t`) a dlouhé (`--tcp`) typy argumentů. V případě pokud nebyl zadán argument `.i`, tak se vypíší dostupné rozhraní pomocí funkce `print_interface()`. Následeně proběhne kontrola pomocí řady podmínek zda byly zadáný nějaké specifické typy paketů, které se mají odchytávat. Tyto argumenty se pomocí funkce `strcat()` spojí do jednoho řetězce a na konci se odstraní nadbytečné slovo `or`. Poté získáme masku a IP adresu sítě pomocí funkce `pcap_lookupnet()` a dále použijeme funkci `pcap_open_live()`, která otevře rozhraní a vrátí popisovač. Pomocí funkce `pcap_compile()` zkompilujeme filtry a pomocí funkce `pcap_setfilter()` tyto filtry nastavíme. Pokud všechny předchoží funkce proběhly bez problému vykonáme funkci `pcap_loop()`, která zpracovává pakety ze živého vysílaní podle nastaveného filtru. Funkci lze předat počet, který má zachytit, který bereme z argumentu `-n` (defaultě je nastavený na 1). Pro zpracování jednotlivých odchycených paketů je této funcki předávána funkce `process_packet()`. Na konci je zařízení uzavřeno pomoc `pcap_close()` a program úspěšně ukončen.
### Implementace funkce process_packet():
V této funkci se vypíše čas, kdy byl paket zachycený pomocí funkce `print_time()`. Jsou vytvořeny struktury pro IP hlavičku a pro jednotlivé pakety protokolů, které obsahují potřebné informace. Pomocí proměnné `ehternet_header`, která se nachází v struktuře `ether_header` se zkontroluje zda se jedná o IPv4, IPv6 nebo ARP. Uvnitř podmínky pro IPv4 se nachází switch, který rozhoduje, ke kterému protokolu paket patří. Vypíší se potřebné informace o paketech a těmi jsou: zdrojová a cílová MAC adresa, délka paketu, zdrojová a cílová IP adresa a zdrojový a cílový port. Pro výpis paketů se používá funkce `print_packet()`. IPv6 funguje podobný způsobem akorát na principu podmínek namísto switche a navíc používá funcki `print_ipv6` pro výpis IPv6 adresy. Porty jsou vypisovány pouze u protokolů TCP a UDP. 

## Testy:
### TCP
### Wireshark:
`telnet google.com`
`./ipk-sniffer -i enp0s3 -p 23 --tcp -n 2`
```sh
# Packet 4 from /var/tmp/wireshark_enp0s3GrexnE.pcapng
- 5
- 2023-04-17 15:39:15,441880221
- 0.000478251
- 10.0.2.15
- 142.251.36.142
- TCP
- 74
- 46696 → 23 [SYN] Seq=0 Win=64240 Len=0 MSS=1460 SACK_PERM=1 TSval=111514150 TSecr=0 WS=128

0000   52 54 00 12 35 02 08 00 27 78 63 87 08 00 45 10   RT..5...'xc...E.
0010   00 3c 50 27 40 00 40 06 2a ed 0a 00 02 0f 8e fb   .<P'@.@.*.......
0020   24 8e b6 68 00 17 27 c1 aa 7e 00 00 00 00 a0 02   $..h..'..~......
0030   fa f0 bf c6 00 00 02 04 05 b4 04 02 08 0a 06 a5   ................
0040   92 26 00 00 00 00 01 03 03 07                     .&........
----
# Packet 5 from /var/tmp/wireshark_enp0s3GrexnE.pcapng
- 6
- 2023-04-17 15:39:16,463080851
- 1.021200630
- 10.0.2.15
- 142.251.36.142
- TCP
- 74
- [TCP Retransmission] [TCP Port numbers reused] 46696 → 23 [SYN] Seq=0 Win=64240 Len=0 MSS=1460 SACK_PERM=1 TSval=111515171 TSecr=0 WS=128

0000   52 54 00 12 35 02 08 00 27 78 63 87 08 00 45 10   RT..5...'xc...E.
0010   00 3c 50 28 40 00 40 06 2a ec 0a 00 02 0f 8e fb   .<P(@.@.*.......
0020   24 8e b6 68 00 17 27 c1 aa 7e 00 00 00 00 a0 02   $..h..'..~......
0030   fa f0 bf c6 00 00 02 04 05 b4 04 02 08 0a 06 a5   ................
0040   96 23 00 00 00 00 01 03 03 07                     .#........
```
### My output:
```
[michal@fedora ipk-sniffer]$ sudo ./ipk-sniffer -i enp0s3 -p 23 --tcp -n 2
timestamp: 2023-04-17T15:39:16.463116+02:00
src MAC: 08:00:27:78:63:87
dst MAC: 52:54:00:12:35:02
frame length: 74 bytes
src IP: 10.0.2.15
dst IP: 142.251.36.142
src port: 46696
dst port: 23

0x0000   52 54 00 12 35 02 08 00  27 78 63 87 08 00 45 10    RT..5...'xc...E.
0x0010   00 3c 50 27 40 00 40 06  2a ed 0a 00 02 0f 8e fb    .<P'@.@.*.......
0x0020   24 8e b6 68 00 17 27 c1  aa 7e 00 00 00 00 a0 02    $..h..'..~......
0x0030   fa f0 bf c6 00 00 02 04  05 b4 04 02 08 0a 06 a5    ................
0x0040   92 26 00 00 00 00 01 03  03 07                      .&........

timestamp: 2023-04-17T15:39:16.463180+02:00
src MAC: 08:00:27:78:63:87
dst MAC: 52:54:00:12:35:02
frame length: 74 bytes
src IP: 10.0.2.15
dst IP: 142.251.36.142
src port: 46696
dst port: 23

0x0000   52 54 00 12 35 02 08 00  27 78 63 87 08 00 45 10    RT..5...'xc...E.
0x0010   00 3c 50 28 40 00 40 06  2a ec 0a 00 02 0f 8e fb    .<P(@.@.*.......
0x0020   24 8e b6 68 00 17 27 c1  aa 7e 00 00 00 00 a0 02    $..h..'..~......
0x0030   fa f0 bf c6 00 00 02 04  05 b4 04 02 08 0a 06 a5    ................
0x0040   96 23 00 00 00 00 01 03  03 07                      .#........
```
### UDP
### Wireshark
`telnet google.com`
`./ipk-sniffer -i enp0s3 --udp`
```sh
# Packet 6 from /var/tmp/wireshark_enp0s3ko0EmK.pcapng
- 1
- 2023-04-17 15:34:32,483286890
- 0.000000000
- 10.0.2.15
- 192.168.0.1
- DNS
- 70
- Standard query 0xaf09 AAAA google.com
0000   52 54 00 12 35 02 08 00 27 78 63 87 08 00 45 00   RT..5...'xc...E.
0010   00 38 7d 22 00 00 40 11 30 db 0a 00 02 0f c0 a8   .8}"..@.0.......
0020   00 01 a6 f3 00 35 00 24 cc ed af 09 01 00 00 01   .....5.$........
0030   00 00 00 00 00 00 06 67 6f 6f 67 6c 65 03 63 6f   .......google.co
0040   6d 00 00 1c 00 01                                 m.....
```
### My output:
```
[michal@fedora ipk-sniffer]$ sudo ./ipk-sniffer -i enp0s3 --udp
timestamp: 2023-04-17T15:34:33.070745+02:00
src MAC: 08:00:27:78:63:87
dst MAC: 52:54:00:12:35:02
frame length: 70 bytes
src IP: 10.0.2.15
dst IP: 192.168.0.1

0x0000   52 54 00 12 35 02 08 00  27 78 63 87 08 00 45 00    RT..5...'xc...E.
0x0010   00 38 7d 22 00 00 40 11  30 db 0a 00 02 0f c0 a8    .8}"..@.0.......
0x0020   00 01 a6 f3 00 35 00 24  cc ed af 09 01 00 00 01    .....5.$........
0x0030   00 00 00 00 00 00 06 67  6f 6f 67 6c 65 03 63 6f    .......google.co
0x0040   6d 00 00 1c 00 01                                   m.....
```
### ICMP4 a ICMP6 
### Wireshark:
`ping6 ff02::1%enp0s3`
`./ipk-sniffer -i enp0s3 --icmp4 --icmp6`
```

# Packet 0 from /var/tmp/wireshark_enp0s3RpsF3h.pcapng
- 1
- 2023-04-17 15:43:21,629182453
- 0.000000000
- fe80::cee0:4ddc:f719:6824
- ff02::1
- ICMPv6
- 118
- Echo (ping) request id=0x0003, seq=1, hop limit=1 (multicast)

0000   33 33 00 00 00 01 08 00 27 78 63 87 86 dd 60 08   33......'xc...`.
0010   0f d3 00 40 3a 01 fe 80 00 00 00 00 00 00 ce e0   ...@:...........
0020   4d dc f7 19 68 24 ff 02 00 00 00 00 00 00 00 00   M...h$..........
0030   00 00 00 00 00 01 80 00 70 e3 00 03 00 01 f9 4c   ........p......L
0040   3d 64 00 00 00 00 96 99 09 00 00 00 00 00 10 11   =d..............
0050   12 13 14 15 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21   .............. !
0060   22 23 24 25 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31   "#$%&'()*+,-./01
0070   32 33 34 35 36 37                                 234567
```
### My output:
```
[michal@fedora ipk-sniffer]$ sudo ./ipk-sniffer -i enp0s3 --icmp4 --icmp6
timestamp: 2023-04-17T15:43:22.607887+02:00
src MAC: 08:00:27:78:63:87
dst MAC: 33:33:00:00:00:01
frame length: 118 bytes
fe80:0000:0000:0000:cee0:4ddc:f719:6824 > ff02:0000:0000:0000:0000:0000:0000:0001
0x0000   33 33 00 00 00 01 08 00  27 78 63 87 86 dd 60 08    33......'xc...`.
0x0010   0f d3 00 40 3a 01 fe 80  00 00 00 00 00 00 ce e0    ...@:...........
0x0020   4d dc f7 19 68 24 ff 02  00 00 00 00 00 00 00 00    M...h$..........
0x0030   00 00 00 00 00 01 80 00  70 e3 00 03 00 01 f9 4c    ........p......L
0x0040   3d 64 00 00 00 00 96 99  09 00 00 00 00 00 10 11    =d..............
0x0050   12 13 14 15 16 17 18 19  1a 1b 1c 1d 1e 1f 20 21    .............. !
0x0060   22 23 24 25 26 27 28 29  2a 2b 2c 2d 2e 2f 30 31    "#$%&'()*+,-./01
0x0070   32 33 34 35 36 37                                   234567
```
### ARP 
### Wireshark:
`telnet google.com`
`./ipk-sniffer -i enp0s3 --arp`
```
# Packet 4 from /var/tmp/wireshark_enp0s3NfDLUw.pcapng
- 5
- 2023-04-17 15:49:31,438754053
- 5.055423427
- PcsCompu_78:63:87
- RealtekU_12:35:02
- ARP
- 42
- Who has 10.0.2.2? Tell 10.0.2.15
0000   52 54 00 12 35 02 08 00 27 78 63 87 08 06 00 01   RT..5...'xc.....
0010   08 00 06 04 00 01 08 00 27 78 63 87 0a 00 02 0f   ........'xc.....
0020   00 00 00 00 00 00 0a 00 02 02                     ..........
```
### My output:
```
[michal@fedora ipk-sniffer]$ sudo ./ipk-sniffer -i enp0s3 --arp
timestamp: 2023-04-17T15:49:31.631364+02:00
src MAC: 08:00:27:78:63:87
dst MAC: 52:54:00:12:35:02
frame length: 42 bytes
src IP: 10.0.2.15
dst IP: 10.0.2.2

0x0000   52 54 00 12 35 02 08 00  27 78 63 87 08 06 00 01    RT..5...'xc.....
0x0010   08 00 06 04 00 01 08 00  27 78 63 87 0a 00 02 0f    ........'xc.....
0x0020   00 00 00 00 00 00 0a 00  02 02                      ..........
```
## ARP 
`sudo tcpreplay -t -i enp0s3 arp_pcap.pcapng.cap`
`./ipk-sniffer -i enp0s3 --arp`
### Wireshark:
```
# Packet 9 from /var/tmp/wireshark_enp0s3DstELy.pcapng
- 10
- 2023-04-17 16:06:52,312568229
- 0.000002757
- c4:01:32:58:00:00
- c4:02:32:6b:00:00
- ARP
- 60
- Who has 10.0.0.2? Tell 10.0.0.1

0000   c4 02 32 6b 00 00 c4 01 32 58 00 00 08 06 00 01   ..2k....2X......
0010   08 00 06 04 00 01 c4 01 32 58 00 00 0a 00 00 01   ........2X......
0020   c4 02 32 6b 00 00 0a 00 00 02 00 00 00 00 00 00   ..2k............
0030   00 00 00 00 00 00 00 00 00 00 00 00               ............
```
### My output:
```
[michal@fedora ipk-sniffer]$ sudo ./ipk-sniffer -i enp0s3 --arp
timestamp: 2023-04-17T16:06:52.719033+02:00
src MAC: c4:01:32:58:00:00
dst MAC: c4:02:32:6b:00:00
frame length: 60 bytes
src IP: 10.0.0.1
dst IP: 10.0.0.2

0x0000   c4 02 32 6b 00 00 c4 01  32 58 00 00 08 06 00 01    ..2k....2X......
0x0010   08 00 06 04 00 01 c4 01  32 58 00 00 0a 00 00 01    ........2X......
0x0020   c4 02 32 6b 00 00 0a 00  00 02 00 00 00 00 00 00    ..2k............
0x0030   00 00 00 00 00 00 00 00  00 00 00 00                ............

CHANGELOG.md
# Changelog
- 17.04.2023
	- Fixed CHANGELOG.md
	- Added CHANGELOG.md
	- fix/readme
	- Merge pull request 'Added comments and readme' (#7) from feature/comments_and_readme into master
	- Added comments and readme
	- Merge pull request 'Added arp and ipv6.' (#6) from feature/arp_and_ipv6 into master
	- Added arp and ipv6
	- Added timestamp
	- Added timestamp
- 16.04.2023
	- Added arp and ipv6.
	- Merge pull request 'Added printing and filters' (#3) from feature/print_protocols into master
	- Added printing and filters
- 15.04.2023
	- Merge pull request 'Added processPacket' (#2) from feature/processPacket into master
	- Added processPacket
	- Merge pull request 'Added makefile and basic structure' (#1) from feature/structure_and_makefile into master
	- Added makefile and basic structure
	- Initial commit


ipk-sniffer.c
#include <stdio.h>
#include <stdlib.h>
#include <pcap.h>
#include <getopt.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <net/ethernet.h>
#include <netinet/if_ether.h>
#include <netinet/ip_icmp.h>
#include <netinet/icmp6.h>
#include <netinet/ip6.h>
#include <netinet/in.h>
#include <net/if.h>
#include <arpa/inet.h>
#include <time.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>

#define BUFFER_SIZE 65536

/* 
Global variables for flags and interface
Flag set to 1 needs to be sniffed.
*/
int interface_flag = 0;
char interface[50];
int tcp_flag = 0;
int udp_flag = 0;
int port = -1;
int icmp4_flag = 0;
int icmp6_flag = 0;
int arp_flag = 0;
int ndp_flag = 0;
int igmp_flag = 0;
int mld_flag = 0;
int num_flag = 0;
int num = 1;

/* 
Function to print all interfaces that are available.
At the end it will use pcap_freealldevs() function to free list of devices that allocated pcap_findalldevs().
*/
void print_interface(pcap_if_t *interfaces);

/* 
Function to print ipv6 adress.
Function inspired by: https://stackoverflow.com/questions/3727421/expand-an-ipv6-address-so-i-can-print-it-to-stdout
*/
void print_ipv6(const struct in6_addr *addr);

/* 
Function to print ethernet header.
Prints src MAC, dst MAC and frame length.
*/
void print_eth_hdr(const u_char *packet, int len);

/* 
Function to print data in rows of 16 bytes: offset   hex   ascii.
Function inspired by: https://www.tcpdump.org/other/sniffex.c
*/
void print_hex_ascii_line(const u_char *payload, int len, int offset);

/* 
Function to print timestamp.
Function inspired by: https://www.tcpdump.org/other/sniffex.c
and 
https://www.programcreek.com/cpp/?code=mq1n%2FNoMercy%2FNoMercy-master%2FSource%2FClient%2FNM_Engine%2FINetworkScanner.cpp
*/
void print_packet(const u_char *buffer, int len);

/* 
Function to print timestamp.
*/
void print_time(const u_char *addr, int size);

/* 
Function to print usage.
*/
void usage();

/* 
Callback function for pcap_loop(), which handles recieved packets.
Function checks if packet is ipv4 or ipv6 and what protocol.
Prints timestamp of packet, encoded data of packet, packet and ipv6 adress
*/
void process_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet);


void print_ipv6(const struct in6_addr *addr)
{
    printf("%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x:%02x%02x",
           (int)addr->s6_addr[0], (int)addr->s6_addr[1],
           (int)addr->s6_addr[2], (int)addr->s6_addr[3],
           (int)addr->s6_addr[4], (int)addr->s6_addr[5],
           (int)addr->s6_addr[6], (int)addr->s6_addr[7],
           (int)addr->s6_addr[8], (int)addr->s6_addr[9],
           (int)addr->s6_addr[10], (int)addr->s6_addr[11],
           (int)addr->s6_addr[12], (int)addr->s6_addr[13],
           (int)addr->s6_addr[14], (int)addr->s6_addr[15]);
}


void print_time(const u_char *addr, int size)
{
    struct timeval tv;
    time_t nowtime;
    struct tm *nowtm;

    gettimeofday(&tv, NULL);
    nowtime = tv.tv_sec;
    nowtm = localtime(&nowtime);

    char timestr[64];
    strftime(timestr, sizeof(timestr), "%FT%T", nowtm);
    printf("timestamp: %s.%06ld+02:00\n", timestr, tv.tv_usec);
}


void print_hex_ascii_line(const u_char *payload, int len, int offset)
{
    int i;
    int gap;
    const u_char *ch;

    /* offset */
    printf("0x%04x   ", offset);

    /* hex */
    ch = payload;
    for (i = 0; i < len; i++)
    {
        printf("%02x ", *ch);
        ch++;
        /* print extra space after 8th byte for visual aid */
        if (i == 7)
            printf(" ");
    }

    /* print space to handle line less than 8 bytes */
    if (len < 8)
        printf(" ");

    /* fill hex gap with spaces if not full line */
    if (len < 16)
    {
        gap = 16 - len;
        for (i = 0; i < gap; i++)
        {
            printf("   ");
        }
    }

    printf("   ");
    /* ascii (if printable) */
    ch = payload;

    for (i = 0; i < len; i++)
    {
        if (isprint(*ch))
            printf("%c", *ch);
        else
            printf(".");
        ch++;
    }
    printf("\n");

    return;
}

void print_packet(const u_char *buffer, int len)
{

    int len_rem = len;
    int line_width = 16; /* number of bytes per line */
    int line_len;
    int offset = 0; /* zero-based offset counter */
    const u_char *buff = buffer;

    if (len <= 0)
        return;

    /* data fits on one line */
    if (len <= line_width)
    {
        print_hex_ascii_line(buff, len, offset);
        return;
    }

    /* data spans multiple lines */
    for (;;)
    {
        /* compute current line length */
        line_len = line_width % len_rem;
        /* compute total remaining */
        len_rem -= line_len;
        /* print line */
        print_hex_ascii_line(buff, line_len, offset);
        /* shift pointer to remaining bytes to print */
        buff += line_len;
        /* add offset */
        offset += line_width;
        /* check if we have line width chars or less */
        if (len_rem <= line_width)
        {
            /* print last line and get out */
            print_hex_ascii_line(buff, len_rem, offset);
            offset += line_width;
            break;
        }
    }
    printf("\n");
    return;
}

void process_packet(u_char *args, const struct pcap_pkthdr *header, const u_char *packet)
{
    int size = header->len;
    const struct iphdr *ip_header = (struct iphdr *)(packet + sizeof(struct ethhdr));
    u_short ip_len = (ip_header->ihl) * 4;

    const struct ether_header *ethernet_header = (struct ether_header *)packet;

    struct ether_arp *arp = (struct ether_arp *)(packet + sizeof(struct ether_header) + ip_len);
    const struct tcphdr *tcp_header = (struct tcphdr *)(packet + sizeof(struct ether_header) + ip_len);
    const struct ip6_hdr *ip6_header;
    const struct udphdr *udp_header = (struct udphdr *)(packet + sizeof(struct ether_header) + ip_len);
    struct ip *ip = (struct ip *)(packet + sizeof(struct ether_header));

    print_time(packet, size);

    if (ntohs(ethernet_header->ether_type) == ETHERTYPE_IP)
    { // if ETHERTYPE is IPV4

        switch (ip_header->protocol) // Check the Protocol and do accordingly...
        {
        case 1: // ICMPv4 IPV4
            print_eth_hdr(packet, size);
            printf("src IP: %s\n", inet_ntoa(ip->ip_src));
            printf("dst IP: %s\n", inet_ntoa(ip->ip_dst));
            printf("\n");
            print_packet(packet, size);
            break;

        case 2: // IGMP
            print_eth_hdr(packet, size);
            printf("src IP: %s\n", inet_ntoa(ip->ip_src));
            printf("dst IP: %s\n", inet_ntoa(ip->ip_dst));
            printf("\n");
            print_packet(packet, size);
            break;

        case 6: // TCP IPV4
            print_eth_hdr(packet, size);
            printf("src IP: %s\n", inet_ntoa(ip->ip_src));
            printf("dst IP: %s\n", inet_ntoa(ip->ip_dst));
            if (port != -1)
            {
                printf("src port: %d\n", ntohs(tcp_header->th_sport));
                printf("dst port: %d\n", ntohs(tcp_header->th_dport));
            }
            printf("\n");
            print_packet(packet, size);
            break;
        case 17: // UDP IPV4
            print_eth_hdr(packet, size);
            printf("src IP: %s\n", inet_ntoa(ip->ip_src));
            printf("dst IP: %s\n", inet_ntoa(ip->ip_dst));
            if (port != -1)
            {
                printf("src port: %d\n", ntohs(udp_header->uh_dport));
                printf("dst port: %d\n", ntohs(udp_header->uh_dport));
            }
            printf("\n");
            print_packet(packet, size);
            break;

        default:
            break;
        }
    }
    else if (ntohs(ethernet_header->ether_type) == ETHERTYPE_IPV6)
    { // if ETHERTYPE is IPV6
        ip6_header = (struct ip6_hdr *)(packet + sizeof(struct ethhdr));
        int protocol = ip6_header->ip6_nxt;

        if (protocol == 6)
        { // TCP IPV6
            print_eth_hdr(packet, size);
            print_ipv6(&ip6_header->ip6_src);
            printf(" > ");
            print_ipv6(&ip6_header->ip6_dst);
            if (port != -1)
            {
                printf("src port: %d\n", ntohs(tcp_header->th_sport));
                printf("dst port: %d\n", ntohs(tcp_header->th_dport));
            }
            printf("\n");
            print_packet(packet, size);
        }
        else if (protocol == 17)
        { // UDP IPV6
            print_eth_hdr(packet, size);
            print_ipv6(&ip6_header->ip6_src);
            printf(" > ");
            print_ipv6(&ip6_header->ip6_dst);
            if (port != -1)
            {
                printf("src port: %d\n", ntohs(udp_header->uh_sport));
                printf("dst port: %d\n", ntohs(udp_header->uh_dport));
            }
            printf("\n");
            print_packet(packet, size);
        }
        else if (protocol == 58)
        { // ICMPv6 IPV6
            print_eth_hdr(packet, size);
            print_ipv6(&ip6_header->ip6_src);
            printf(" > ");
            print_ipv6(&ip6_header->ip6_dst);
            printf("\n");
            print_packet(packet, size);
        }
        else if (protocol == 143)
        { // MLD IPV6
            print_eth_hdr(packet, size);
            print_ipv6(&ip6_header->ip6_src);
            printf(" > ");
            print_ipv6(&ip6_header->ip6_dst);
            printf("\n");
            print_packet(packet, size);
        }
    }
    else if (ntohs(ethernet_header->ether_type) == ETHERTYPE_ARP)
    { // ARP
        char srcIP[16], destIP[16];
        inet_ntop(AF_INET, &(arp->arp_spa), srcIP, sizeof(srcIP));
        inet_ntop(AF_INET, &(arp->arp_tpa), destIP, sizeof(destIP));
        print_eth_hdr(packet, size);
        printf("src IP: %s\n", srcIP);
        printf("dst IP: %s\n", destIP);
        printf("\n");
        print_packet(packet, size);
    }
}

void usage()
{
    printf("Usage: ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] [--ndp] {-n num}\n");
}

void print_eth_hdr(const u_char *packet, int len)
{
    const struct ether_header *ethernet_header = (struct ether_header *)packet;
    printf("src MAC: %02x:%02x:%02x:%02x:%02x:%02x\n", ethernet_header->ether_shost[0], ethernet_header->ether_shost[1], ethernet_header->ether_shost[2], ethernet_header->ether_shost[3], ethernet_header->ether_shost[4], ethernet_header->ether_shost[5]);
    printf("dst MAC: %02x:%02x:%02x:%02x:%02x:%02x\n", ethernet_header->ether_dhost[0], ethernet_header->ether_dhost[1], ethernet_header->ether_dhost[2], ethernet_header->ether_dhost[3], ethernet_header->ether_dhost[4], ethernet_header->ether_dhost[5]);
    printf("frame length: %d bytes\n", len);
}

void print_interface(pcap_if_t *interfaces)
{
    while (interfaces->next != NULL)
    {
        printf("Name: %s   Description: %s \n", interfaces->name, interfaces->description);
        interfaces = interfaces->next;
    }
    pcap_freealldevs(interfaces);
    exit(0);
}

int main(int argc, char *argv[])
{
    char errbuf[PCAP_ERRBUF_SIZE];
    pcap_t *handle;
    struct bpf_program fp;
    char filter_exp[BUFFER_SIZE];
    bpf_u_int32 mask;
    bpf_u_int32 net;
    const u_char *packet;
    int opt;
    char port_str[50];
    pcap_if_t *interfaces;

    /* 
     Get a list of capture devices.
    */
    if (pcap_findalldevs(&interfaces, errbuf) == -1)
    {
        fprintf(stderr, "ERROR: No interfaces available. INFO: %s\n", errbuf);
        return EXIT_FAILURE;
    }

    /* 
     Argument wasnt set -> print interfaces.
    */
    if (argc < 2) 
    {
        print_interface(interfaces);
    }

    /* 
     While loop that checks what arguments were given and sets flags.
     Checks if arguments are valid.
    */
    while ((opt = getopt(argc, argv, "i:p:tun:-:")) != -1)
    {
        switch (opt)
        {
        case 'i': // interface
            if(interface_flag == 1){
                fprintf(stderr, "ERROR: Interface already set. INFO: %s\n", errbuf);
                return EXIT_FAILURE;
            }
            interface_flag = 1;
            if (optarg[0] == '-') //check if there is name and not argument after -i
            {
                print_interface(interfaces); // in case of wrong arguments print interface
            }
            strncpy(interface, optarg, 50);
            break;
        case 'p': // port
            port = atoi(optarg);
            sprintf(port_str, "%d", port);
            if (port <= 0 || port >= 65536) //check range of port
            {
                fprintf(stderr, "Invalid port number: %d\n", port);
                exit(EXIT_FAILURE);
            }
            break;
        case 't': //tcp
            tcp_flag = 1;
            break;
        case 'u': //udp
            udp_flag = 1;
            break;
        case 'n': // number of packets
            num_flag = 1;
            num = atoi(optarg);
            break;
        case '-':
            if (strcmp(optarg, "tcp") == 0)
            {
                tcp_flag = 1;
            }
            else if (strcmp(optarg, "udp") == 0)
            {
                udp_flag = 1;
            }
            else if (strcmp(optarg, "icmp4") == 0)
            {
                icmp4_flag = 1;
            }
            else if (strcmp(optarg, "icmp6") == 0)
            {
                icmp6_flag = 1;
            }
            else if (strcmp(optarg, "arp") == 0)
            {
                arp_flag = 1;
            }
            else if (strcmp(optarg, "ndp") == 0)
            {
                ndp_flag = 1;
            }
            else if (strcmp(optarg, "igmp") == 0)
            {
                igmp_flag = 1;
            }
            else if (strcmp(optarg, "mld") == 0)
            {
                mld_flag = 1;
            }
            else if (strcmp(optarg, "interface") == 0)
            {
                if(interface_flag == 1){
                    fprintf(stderr, "ERROR: Interface already set. INFO: %s\n", errbuf);
                    return EXIT_FAILURE;
                }
                interface_flag = 1;
                if (argv[optind] && argv[optind][0] != '-') //check if there is name and not argument after -i
                {
                    strncpy(interface, argv[optind], 50);
                    optind++;
                }
                else // in case of wrong arguments print interface
                {
                    print_interface(interfaces);
                }
            }
            break;
        }
    }

    if (interface_flag == 0) // interface not set so print interfaces
    {
        print_interface(interfaces);
    }

    if (tcp_flag == 1) // if tcp argument is given
    {
        if (port != -1) // concatenate tcp with port
        {    
            strcat(filter_exp, "tcp port ");
            strcat(filter_exp, port_str);
            strcat(filter_exp, " or "); // example: tcp port 23
        }
        else
        {
            strcat(filter_exp, "tcp or "); // concatenate only tcp
        }
    }
    if (udp_flag == 1) // if udp argument is given
    {
        if (port != -1) // concatenate udp with port
        {
            strcat(filter_exp, "udp port ");
            strcat(filter_exp, port_str);
            strcat(filter_exp, " or "); // example: udp port 23
        }
        else
        {
            strcat(filter_exp, "udp or "); // concatenate only udp
        }
    }
    if (icmp4_flag == 1) // if icmp4 argument is given
    {
        strcat(filter_exp, "icmp or "); // concatenate only icmp4
    }
    if (icmp6_flag == 1)
    {
        strcat(filter_exp, "icmp6 or ");
    }
    if (arp_flag == 1)
    {
        strcat(filter_exp, "arp or ");
    }
    if (igmp_flag == 1)
    {
        strcat(filter_exp, "igmp or ");
    }
    if (mld_flag == 1)
    {
        strcat(filter_exp, "mld or ");
    }

    filter_exp[strlen(filter_exp) - 3] = '\0'; // Remove redundant 'or'

    /* Find the properties for the device */
    if (pcap_lookupnet(interface, &net, &mask, errbuf) == -1)
    {
        fprintf(stderr, "Couldn't get netmask for device %s: %s\n", interface, errbuf);
        net = 0;
        mask = 0;
    }

    /* Open the session in promiscuous mode */
    handle = pcap_open_live(interface, BUFFER_SIZE, 1, 1000, errbuf);
    if (handle == NULL)
    {
        fprintf(stderr, "Couldn't open device %s: %s\n", interface, errbuf);
        exit(1);
    }

    /* Compile and apply the filter */
    if (pcap_compile(handle, &fp, filter_exp, 0, net) == -1)
    {
        fprintf(stderr, "Couldn't parse filter %s: %s\n", filter_exp, pcap_geterr(handle));
        return (2);
    }
    if (pcap_setfilter(handle, &fp) == -1)
    {
        fprintf(stderr, "Couldn't install filter %s: %s\n", filter_exp, pcap_geterr(handle));
        return (2);
    }
    /* process packets from a live capture or savefile */
    pcap_loop(handle, num, process_packet, NULL);

    pcap_close(handle);

    exit(0);
}